### ISSUE TITLE
DNS resolution doesn't work 

### ISSUE THREAD

**Issue by NikitaSkrynnik** [projectcalico/vpp-dataplane#584]
**Environment**
- Calico/VPP version: v3.25
- Kubernetes version: 1.25.3
- Deployment type: kind cluster
- Network configuration: Calico with default settings. VPP dataplane

**Issue description**
It looks like external DNS requests doesn't work. I created kind cluster with Calico VPP following this [guide](https://github.com/projectcalico/vpp-dataplane/blob/master/docs/developper_guide.md).
I deployed `dnsutils` pod from this [guide](https://kubernetes.io/docs/tasks/administer-cluster/dns-debugging-resolution/#:~:text=kubectl%20apply%20%2Df%20https%3A//k8s.io/examples/admin/dns/dnsutils.yaml) and tried `nslookup www.google.com` and it didn't work. However, `nslookup kubernetes.default` worked.

**To Reproduce**
1. Setup kind cluster using this [guide](https://github.com/projectcalico/vpp-dataplane/blob/master/docs/developper_guide.md)
2. Apply `dnsutils` pod - `kubectl apply -f https://k8s.io/examples/admin/dns/dnsutils.yaml`
3. Do `nslookup www.google.com` from `dnsutils` pod

**Expected behavior**
DNS requests for external resources work correctly.

**Additional context**
1. `apk add` command in `alpine` pod also doesn't work.
2. `nslookup www.google.com 8.8.8.8` works correctly.

**Comment by NikitaSkrynnik** [comment]
@sknat Could you please take a look?

**Comment by onong** [comment]
Hi @NikitaSkrynnik , could you share the vpp-manager logs, ie,

kubectl logs <calico-vpp-node-xyzabc> -n calico-vpp-dataplane -c vpp

**Comment by anaperic** [comment]
If it helps, we are seeing exactly this issue too. We are not running kind, but cluster API orchestrated k8s cluster.
I think we have a race condition it seems where my suspicion is that calico-vpp node comes up before we even manage to get proper felixconfig. 
Because of this, none of the pods sitting on that first control plane node are just not getting out.
This also has of course repricusion on core-dns that can't reach the configured DNS at all, which also means DNS inside the cluster is not working either.

We also did not solve this issue, what is very apparent that if one rolls the node we do not hit the race condition and things start working.

I believe there was a change in calico-vpp 3.25 compared to 3.24.1 behaviour we saw, and for our use-case this is also something that needs urgent fixing.

**Comment by sknat** [comment]
Thanks for the report @NikitaSkrynnik 
There might be two issues emebeded in this story.

@anaperic I definitely saw a similar issue to the scenario you are describing in some deployments, I'll work on a fix for this and keep you posted.

Additionally, regarding DNS resolution failing in kind, I stumbled on a similar issue caused by a corrupted checksum. This was seemingly addressed by #558 but I still needed to do some checks on this PR before merging it. You can probably check with the branch's images once I manage to get them pushed to the repo, see if it helps.

**Comment by NikitaSkrynnik** [comment]
Hi @onong. Here is a cluster dump
[logs.zip](https://github.com/projectcalico/vpp-dataplane/files/11295352/logs.zip)

Here I deployed `dnsutils` pod and did `nslookup www.google.com`

**Comment by NikitaSkrynnik** [comment]
@sknat Hello, thanks for the answer. I downloaded this branch and tried to setup calico-vpp kind cluster, but it just didn't start. Looking forward to updates on this PR. It's very important for our project.

**Comment by sknat** [comment]
Hi @NikitaSkrynnik 
Sorry for the late update, I think I missed a few things in this PR, I'll try to update it in the coming days and keep you posted

**Comment by sknat** [comment]
Hi @NikitaSkrynnik 
Finally managed to get the build of #558 to complete. You should be able to try with :
````
calicovpp/vpp/223d4262d5cca8fa874139bc939eb0bbf8495eb7
calicovpp/agent223d4262d5cca8fa874139bc939eb0bbf8495eb7
````

**Comment by NikitaSkrynnik** [comment]
@sknat I managed to setup a kind cluster with new images, but `csi-node-driver` pods constantly crash and DNS still doesn't work:

Pods:
```
NAMESPACE              NAME                                         READY   STATUS             RESTARTS        AGE
calico-apiserver       calico-apiserver-7c64cc7b77-b5944            1/1     Running            0               6m34s
calico-apiserver       calico-apiserver-7c64cc7b77-pbw8j            1/1     Running            0               6m34s
calico-system          calico-kube-controllers-5cd75bfc95-grrxs     1/1     Running            0               18m
calico-system          calico-node-kppng                            1/1     Running            0               18m
calico-system          calico-node-m4bwt                            1/1     Running            0               18m
calico-system          calico-node-m988s                            1/1     Running            0               18m
calico-system          calico-node-rwn9t                            1/1     Running            0               18m
calico-system          calico-typha-6df9cf7b7d-xkmh8                1/1     Running            0               18m
calico-system          calico-typha-6df9cf7b7d-xqthr                1/1     Running            0               18m
calico-system          csi-node-driver-kzfgm                        1/2     CrashLoopBackOff   6 (28s ago)     18m
calico-system          csi-node-driver-m5vhf                        1/2     CrashLoopBackOff   5 (2m13s ago)   18m
calico-system          csi-node-driver-sbjng                        1/2     CrashLoopBackOff   5 (2m5s ago)    18m
calico-system          csi-node-driver-xzfsz                        1/2     CrashLoopBackOff   6 (41s ago)     18m
calico-vpp-dataplane   calico-vpp-node-2prn8                        2/2     Running            0               17m
calico-vpp-dataplane   calico-vpp-node-2xqgh                        2/2     Running            0               17m
calico-vpp-dataplane   calico-vpp-node-6q8rk                        2/2     Running            0               17m
calico-vpp-dataplane   calico-vpp-node-b74h2                        2/2     Running            0               17m
default                dnsutils                                     1/1     Running            0               4m30s
kube-system            coredns-565d847f94-64hh4                     1/1     Running            0               21m
kube-system            coredns-565d847f94-t9cvf                     1/1     Running            0               21m
kube-system            etcd-kind-control-plane                      1/1     Running            0               21m
kube-system            kube-apiserver-kind-control-plane            1/1     Running            0               21m
kube-system            kube-controller-manager-kind-control-plane   1/1     Running            0               21m
kube-system            kube-proxy-gwg24                             1/1     Running            0               21m
kube-system            kube-proxy-pbsbz                             1/1     Running            0               21m
kube-system            kube-proxy-wz7p9                             1/1     Running            0               21m
kube-system            kube-proxy-zsjs4                             1/1     Running            0               21m
kube-system            kube-scheduler-kind-control-plane            1/1     Running            0               21m
local-path-storage     local-path-provisioner-684f458cdd-5pwmm      1/1     Running            0               21m
tigera-operator        tigera-operator-55dd57cddb-wlmgl             1/1     Running            0               21m
```

calico-vpp-node desribe:
```
Name:           calico-vpp-node
Selector:       k8s-app=calico-vpp-node
Node-Selector:  kubernetes.io/os=linux
Labels:         k8s-app=calico-vpp-node
Annotations:    deprecated.daemonset.template.generation: 2
Desired Number of Nodes Scheduled: 4
Current Number of Nodes Scheduled: 4
Number of Nodes Scheduled with Up-to-date Pods: 4
Number of Nodes Scheduled with Available Pods: 4
Number of Nodes Misscheduled: 0
Pods Status:  4 Running / 0 Waiting / 0 Succeeded / 0 Failed
Pod Template:
  Labels:           k8s-app=calico-vpp-node
  Service Account:  calico-vpp-node-sa
  Containers:
   vpp:
    Image:      docker.io/calicovpp/vpp:223d4262d5cca8fa874139bc939eb0bbf8495eb7
    Port:       <none>
    Host Port:  <none>
    Limits:
      memory:  80Gi
    Requests:
      cpu:     1
      memory:  4Gi
    Environment Variables from:
      calico-vpp-config  ConfigMap  Optional: false
    Environment:
      DATASTORE_TYPE:      kubernetes
      WAIT_FOR_DATASTORE:  true
      NODENAME:             (v1:spec.nodeName)
      LD_LIBRARY_PATH:     /repo/vpp-manager/vpp_build/build-root/install-vpp-native/vpp/
    Mounts:
      /dev from devices (rw)
      /etc/ssl/certs/ from ssl-certs (rw)
      /etc/vpp from vpp-config (rw)
      /host from host-root (rw)
      /lib/firmware from lib-firmware (rw)
      /repo from repo-directory (rw)
      /run/netns/ from netns (rw)
      /sys from hostsys (rw)
      /usr/share/ca-certificates from share-certs (rw)
      /var/lib/vpp from vpp-data (rw)
      /var/run/vpp from vpp-rundir (rw)
   agent:
    Image:      docker.io/calicovpp/agent:223d4262d5cca8fa874139bc939eb0bbf8495eb7
    Port:       <none>
    Host Port:  <none>
    Requests:
      cpu:  250m
    Environment Variables from:
      calico-vpp-config  ConfigMap  Optional: false
    Environment:
      DATASTORE_TYPE:      kubernetes
      WAIT_FOR_DATASTORE:  true
      NODENAME:             (v1:spec.nodeName)
      NAMESPACE:            (v1:metadata.namespace)
    Mounts:
      /run/netns/ from netns (rw)
      /var/lib/calico/felix-plugins from felix-plugins (rw)
      /var/run/calico from var-run-calico (rw)
      /var/run/vpp from vpp-rundir (rw)
  Volumes:
   repo-directory:
    Type:          HostPath (bare host directory volume)
    Path:          /home/nikita/repos/vpp-dataplane
    HostPathType:  
   ssl-certs:
    Type:          HostPath (bare host directory volume)
    Path:          /etc/ssl/certs/
    HostPathType:  
   share-certs:
    Type:          HostPath (bare host directory volume)
    Path:          /usr/share/ca-certificates
    HostPathType:  
   lib-firmware:
    Type:          HostPath (bare host directory volume)
    Path:          /lib/firmware
    HostPathType:  
   vpp-rundir:
    Type:          HostPath (bare host directory volume)
    Path:          /var/run/vpp
    HostPathType:  
   vpp-data:
    Type:          HostPath (bare host directory volume)
    Path:          /var/lib/vpp
    HostPathType:  DirectoryOrCreate
   vpp-config:
    Type:          HostPath (bare host directory volume)
    Path:          /etc/vpp
    HostPathType:  
   devices:
    Type:          HostPath (bare host directory volume)
    Path:          /dev
    HostPathType:  
   hostsys:
    Type:          HostPath (bare host directory volume)
    Path:          /sys
    HostPathType:  
   var-run-calico:
    Type:          HostPath (bare host directory volume)
    Path:          /var/run/calico
    HostPathType:  
   netns:
    Type:          HostPath (bare host directory volume)
    Path:          /run/netns
    HostPathType:  
   felix-plugins:
    Type:          HostPath (bare host directory volume)
    Path:          /var/lib/calico/felix-plugins
    HostPathType:  
   host-root:
    Type:               HostPath (bare host directory volume)
    Path:               /
    HostPathType:       
  Priority Class Name:  system-node-critical
Events:
  Type    Reason            Age   From                  Message
  ----    ------            ----  ----                  -------
  Normal  SuccessfulCreate  19m   daemonset-controller  Created pod: calico-vpp-node-ggzcc
  Normal  SuccessfulCreate  19m   daemonset-controller  Created pod: calico-vpp-node-vnkjg
  Normal  SuccessfulCreate  19m   daemonset-controller  Created pod: calico-vpp-node-pbjhp
  Normal  SuccessfulCreate  19m   daemonset-controller  Created pod: calico-vpp-node-9wxww
  Normal  SuccessfulDelete  18m   daemonset-controller  Deleted pod: calico-vpp-node-9wxww
  Normal  SuccessfulDelete  18m   daemonset-controller  Deleted pod: calico-vpp-node-ggzcc
  Normal  SuccessfulDelete  18m   daemonset-controller  Deleted pod: calico-vpp-node-vnkjg
  Normal  SuccessfulDelete  18m   daemonset-controller  Deleted pod: calico-vpp-node-pbjhp
  Normal  SuccessfulCreate  18m   daemonset-controller  Created pod: calico-vpp-node-b74h2
  Normal  SuccessfulCreate  18m   daemonset-controller  Created pod: calico-vpp-node-6q8rk
  Normal  SuccessfulCreate  18m   daemonset-controller  Created pod: calico-vpp-node-2prn8
  Normal  SuccessfulCreate  18m   daemonset-controller  Created pod: calico-vpp-node-2xqgh
```

Maybe I'm doing something wrong. I used this [guide](https://github.com/projectcalico/vpp-dataplane/blob/master/docs/developper_guide.md) again.

I skipped [this part](https://github.com/projectcalico/vpp-dataplane/blob/master/docs/developper_guide.md#:~:text=Build%20development%20images) and changed docker images [here](https://github.com/projectcalico/vpp-dataplane/blob/master/docs/developper_guide.md#:~:text=export%20CALICO_AGENT_IMAGE%3Dlocalhost%3A5000/calicovpp/agent%3Alatest%0Aexport%20CALICO_VPP_IMAGE%3Dlocalhost%3A5000/calicovpp/vpp%3Alatest%0Aexport%20MULTINET_MONITOR_IMAGE%3Dlocalhost%3A5000/calicovpp/multinet%2Dmonitor%3Alatest)
I set the following images:
```
export CALICO_AGENT_IMAGE=docker.io/calicovpp/agent:223d4262d5cca8fa874139bc939eb0bbf8495eb7
export CALICO_VPP_IMAGE=docker.io/calicovpp/vpp:223d4262d5cca8fa874139bc939eb0bbf8495eb7
export MULTINET_MONITOR_IMAGE=docker.io/calicovpp/multinet-monitor:latest
```

**Comment by denis-tingaikin** [comment]
@sknat , @onong , @anaperic Do you have any updates on this?

**Comment by sknat** [comment]
Hi @denis-tingaikin,
We were looking at this just today, we first thought this to just be a checksum issue, but there seems to be another problem between the way docker does DNS resolution in the pods and the way VPP inserts itself on the packet path in order to provide connectivity to the pods.

From what I understand, kind provides DNS by dNATing packets destined to the bridge gateway to port 53 to a specific port on localhost in the pods. This plays well with Calico/linux as iptables rules just add up.
When we install Calico/VPP though, on the packet path between containers and the uplink (so in the case of kind the bridge veth) we bypass iptables rules. So DNS packets end up on the bridge unNATed and never reach the server.

This should sum up the two packet path :

![kind-DNS-issue drawio](https://github.com/projectcalico/vpp-dataplane/assets/3398628/951fcc3a-52a5-4958-aa6c-b5babde0d5ee)

We're looking at two solutions : Either there is a way to change the way kind sets up the DNS resolution, and have it work through the bridge. We can also add an extra classifier rule in VPP to extra the DNS packets and force them through linux.
We'll keep you posted on what we find.

**Comment by sknat** [comment]
@hedibouattour is looking at a fix for this, #627 should contain a fix for this

**Comment by hedibouattour** [comment]
This should be resolved now, with #627. It is merged.
It can be tested with images `calicovpp/vpp:latest`  and `calicovpp/agent:latest`.

**Comment by sknat** [comment]
@denis-tingaikin as @hedibouattour mentioned the `:latest` images should now contain a fix.
For it to work, you'll also need to patch the kind configuration with
```yaml
  CALICOVPP_INITIAL_CONFIG: |-
    {
      "vppStartupSleepSeconds": 1,
      "corePattern": "/var/lib/vpp/vppcore.%e.%p"
      "redirectToHostRules": [
        {
          "proto": "udp",
          "port": 53,
          "ip": "172.18.0.1"
        },
        {
          "proto": "tcp",
          "port": 53,
          "ip": "172.18.0.1"
        }
      ]
    }
kind: ConfigMap
metadata:
  name: calico-vpp-config
  namespace: calico-vpp-dataplane
```
assuming the bridge GW address used by your kind is `172.18.0.1`

Can you test if this is working your setup ?
